openQA.spec

```diff
@@ -45,15 +45,20 @@
 %else
 %define python_scripts_requires %{nil}
 %endif
-%define common_requires perl(Config::IniFiles) perl(Cpanel::JSON::XS) perl(Data::Dump) perl(Getopt::Long) perl(Minion) perl(Mojolicious) perl(Try::Tiny) perl(Regexp::Common)
+%define assetpack_requires perl(Mojolicious::Plugin::AssetPack) => 1.36, perl(CSS::Minifier::XS) perl(JavaScript::Minifier::XS)
+%define common_requires perl(Config::IniFiles) perl(Cpanel::JSON::XS) perl(Cwd) perl(Data::Dump) perl(Data::Dumper) perl(Digest::MD5) perl(Getopt::Long) perl(Minion) => 9.09, perl(Mojolicious) >= 7.92, perl(Try::Tiny) perl(Regexp::Common)
 # runtime requirements for the main package that are not required by other sub-packages
-%define main_requires perl(Date::Format) perl(DateTime::Format::Pg) perl(DBD::Pg) perl(DBI) >= 1.632, perl(DBIx::Class) => 0.082801, perl(DBIx::Class::DeploymentHandler) perl(DBIx::Class::DynamicDefault) perl(DBIx::Class::Schema::Config) perl(DBIx::Class::Storage::Statistics) perl(DBIx::Class::OptimisticLocking) perl(File::Copy::Recursive) perl(Net::OpenID::Consumer) perl(Module::Pluggable) perl(Mojolicious::Plugin::RenderFile) perl(Mojolicious::Plugin::AssetPack) => 1.36, perl(aliased) perl(Config::Tiny) perl(Text::Diff) perl(CommonMark) perl(JSON::Validator) perl(IPC::Run) perl(Archive::Extract) perl(CSS::Minifier::XS) perl(JavaScript::Minifier::XS) perl(Time::ParseDate) perl(Sort::Versions) perl(BSD::Resource) perl(Pod::POM) perl(Mojo::Pg) perl(Mojo::RabbitMQ::Client) => 0.2, perl(SQL::Translator) perl(YAML::XS) perl(LWP::UserAgent)
-%define client_requires perl(IO::Socket::SSL) >= 2.009, perl(LWP::UserAgent)
-%define worker_requires os-autoinst < 5, perl(Mojo::IOLoop::ReadWriteProcess) > 0.19, perl(Minion::Backend::SQLite) perl(Mojo::SQLite) openQA-client
-# all requirements needed by the tests, do not require on this in individual
-# sub-packages except for the devel package
-%define test_requires %common_requires %main_requires %python_scripts_requires %worker_requires perl(Test::MockModule) perl(Test::Output)
-%define devel_requires %test_requires
+%define main_requires %assetpack_requires git-core perl(Carp::Always) perl(Date::Format) perl(DateTime::Format::Pg) perl(DBD::Pg) perl(DBI) >= 1.632, perl(DBIx::Class) => 0.082801, perl(DBIx::Class::DeploymentHandler) perl(DBIx::Class::DynamicDefault) perl(DBIx::Class::Schema::Config) perl(DBIx::Class::Storage::Statistics) perl(DBIx::Class::OptimisticLocking) perl(File::Copy::Recursive) perl(Net::OpenID::Consumer) perl(Module::Pluggable) perl(aliased) perl(Config::Tiny) perl(Text::Diff) perl(CommonMark) perl(JSON::Validator) perl(IPC::Run) perl(Archive::Extract) perl(Time::ParseDate) perl(Sort::Versions) perl(BSD::Resource) perl(Pod::POM) perl(Mojo::Pg) perl(Mojo::RabbitMQ::Client) => 0.2, perl(SQL::Translator) perl(YAML::XS) perl(LWP::UserAgent)
```

Comment:
I can't say that having all in one line is any readable. Just stop using macros, *especially* if only used once.

Reply:
that's actually a change that was already merged as part of #2295 . I am investigating how to make that more readable and at best also share the list with cpanfile but I consider using the openQA-devel package in the container definition more important and then to make it possible to define new dependencies *and* test as part of the same PR.

Reply:
what you should do IMO:

Create another spec file openQA-devel.spec that build requires nothing, but requires the things you need to require. There can also be subpackages for the things you put in macros now. This spec file then hardcodes some dependencies like git-core and chromedriver and takes the rest out of cpanfile by spec file magic.

You can run shell scripts to generate requires in a lot of different ways - e.g. how OBS used to generate its ruby dependencies:
https://build.opensuse.org/package/view_file/OBS:Server:2.6/obs-server/obs-api-deps.spec?expand=1 (line 41)